core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -136,7 +147,13 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
        */
       blockManager.putIterator(key, values, level, tellMaster = true, effectiveStorageLevel)
       blockManager.get(key) match {
-        case Some(v) => v.data.asInstanceOf[Iterator[T]]
+        case Some(v) =>
+          val iter = v.data.asInstanceOf[Iterator[T]]
+          if (v.readMethod != DataReadMethod.Network) {
+            CompletionIterator[T, Iterator[T]](iter, blockManager.releaseLock(key))
```

Comment:
Then actually here we don't need to do the network check, since here we're getting the block that we literally just put in. E.g. the effective storage level here is `DISK_ONLY`, so we just read it from our local disk, in which case we *always* have to release the lock.